// botbuilder-lg/lib/template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Here is a data model that can easily understanded and used as the context or all kinds of visitors
 * wether it's evalator, static checker, anayler.. etc
 */
class Template {
    /**
     * Creates a new instance of the [Template](xref:botbuilder-lg.Template) class.
     * @param templatename Template name without parameters.
     * @param parameters Parameter list.
     * @param templatebody Template content.
     * @param sourceRange [SourceRange](xref:botbuilder-lg.SourceRange) of template.
     */
    constructor(templatename, parameters, templatebody, sourceRange) {
        this.name = templatename || '';
        this.parameters = parameters || [];
        this.sourceRange = sourceRange;
        this.body = templatebody || '';
    }
    /**
     * Returns a string representing the current [Template](xref:botbuilder-lg.Template) object.
     * @returns A string representing the [Template](xref:botbuilder-lg.Template).
     */
    toString() {
        return `[${this.name}(${this.parameters.join(', ')})]"${this.body}"`;
    }
}
exports.Template = Template;
