// botbuilder-lg/lib/templateImport.js
"use strict";
/**
 * @module botbuilder-lg
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Here is a data model that can help users understand and use the LG import definition in LG files easily.
 */
class TemplateImport {
    /**
     * Creates a new instance of the [TemplateImport](xref:botbuilder-lg.TemplateImport) class.
     * @param description Import description, which is in [].
     * @param id Import id, which is a path, in ().
     * @param sourceRange [SourceRange](xref:botbuilder-lg.SourceRange) of template.
     * @param alias Imports alias.
     */
    constructor(description, id, sourceRange, alias) {
        this.toString = () => {
            let importStr = `[${this.description}](${this.id})`;
            if (this.alias) {
                importStr += ` as ${this.alias}`;
            }
            return importStr;
        };
        this.description = description;
        this.sourceRange = sourceRange;
        this.id = id;
        this.alias = alias;
    }
}
exports.TemplateImport = TemplateImport;
